connmgr/conngating.go

```diff
@@ -0,0 +1,71 @@
+package connmgr
+
+import (
+	"github.com/libp2p/go-libp2p-core/control"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/transport"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// ConnectionGater can be implemented by a type that supports active
+// inbound or outbound connection gating.
+//
+// A ConnectionGater will be consulted during different states in the life-cycle of a connection and
+// the specific gating function that will be called depends on the life-cycle state of the
+// connection. Hence, it is important to implement this interface keeping in mind the specific life-cycle state
+// at which you'd like to gate/block the connection.
+//
+// `InterceptDial` and `InterceptPeerDial` are called when we try an outbound dial.
+//
+// `InterceptAccept` is called after we've accepted an inbound connection from a socket but before we
+// begin upgrading it.
+//
+// `InterceptSecured` is called for inbound and outbound connections after we've negotiated
+// the security protocol to use for the connection.
+//
+// `InterceptUpgraded` is called for inbound and outbound connections after we've finished upgrading
+// a connection to have both security and stream multiplexing.
+//
+// This feature can be used to implement *strict* connection management
+// behaviours, such as hard limiting of connections once a max count has been
+// reached.
+//
+// If you'd like to send a disconnect control message to the remote peer for a gated inbound connection,
+// ONLY `InterceptUpgraded` should reject the connection (with an appropriate disconnect reason).
+// All other methods should allow the connection as we can ONLY open control streams
+// for upgraded connections.
+// Note: There's no point in sending disconnect control messages for outbound connections, so we might
+// as well close them as early in the cycle as possible.
+type ConnectionGater interface {
```

Comment:
I'd consider switching from `allow bool` to a Allow, Deny, Pass system. That way, we can compose multiple connection gaters without having to define new interfaces. This would also allow peers to define a "default" policy.

Reply:
@Stebalien 

> "That way, we can compose multiple connection gaters without having to define new interfaces"

Please can you elaborate on this  ? I was thinking of having a `Compose(gater1, gater2) gater` method so users can compose multiple rules together easily but how does a Allow, Deny, Pass system help there ?

Reply:
* Allow -> Immediately allow the dial
* Deny -> Immediately deny the dial.
* Pass/Continue -> Let the next gate check it. Really, this is "no decision" or "does not match any rules".

Reply:
@Stebalien I am not sure we need to add this additional complexity.

This interface allows users to reject connections based on their lifecycle state. We return true if we want to allow a connection based on the passed info and the current lifecycle state of the connection. No gating function call needs to know anything about any other gating function call. I am not sure what functionality `Pass` will give here/if we need it.

As far as "composing" Gaters in concerned, it's easy to do so with the current implementation by providing a `Compose(gater1, gater2) Gater` that returns a `Gater` implementation that only allows a connection if both the input gaters allow the connection for that lifecycle state.

Reply:
@Stebalien If I'm interpreting your proposal correctly, it boils down to adding a HARD ACCEPT action.

I think that breaks composibility and creates invisible coupling between composed gaters. 

We'd like for gaters to only have a local view of the world; they do not know what downstream gaters they would be repressing with their HARD ACCEPT decision.

Imagine a scenario of two gaters:

1. First one accepts "befriended peers".
2. Second one knows the connection limit of the local router and stops accepting connections at N-1, to avoid tripping the router.

If (1) hard accepts connections, (2) might not be given the opportunity to evaluate exactly at the moment of truth, and we might end up tripping up the router.

For that composite to work, (2) would need to precede (1), hence we're implicitly creating spooky action at a distance (i.e. invisible coupling).

Instead, (1) should allow, and (2) should deny. The final result being a DENY.

---

FTR, there are two popular ways one could implement a middleware pattern:

- recursively, by injecting the `next` middleware, and making the current callee call `next` if they want to allow, or return without recursing if they want to deny. This relies on the last `next` being the actual, gated logic.
- iteratively, where middleware components return true/false, and let the framework iterate.

We are choosing the second pattern because it's simpler, and the first pattern is useful in combination with richer APIs, to enable behaviours like "skip next".

Reply:
You've convinced me.

This boils down to ordering. If we say that connection gaters are ordered, then accept, deny, pass make sense. If we're saying they're not ordered, then I agree, we should apply all of them and only accept the connection if they all allow it.

My concern was that I wanted to be able to implement firewall rules. However, the better solution is to just implement a "firewall" connection gater.

```go
fw := Table(
  Rule(Matches("/ip4/127.0.0.1/**"), ResultAllow),
  Rule(Matches("/ip4/6.6.6.6/**"), ResultDeny),
  Rule(Matches(...), Delegate(arbitraryGater)),
  Rule(Matches("/ip6/6.6.6.6/**"), ResultDeny),
  Always(ResultDeny),
)
```